File: packages/expo-updates/ios/EXUpdates/EXUpdatesModule.m
// Copyright 2019 650 Industries. All rights reserved.

#import <EXUpdates/EXUpdatesConfig.h>
#import <EXUpdates/EXUpdatesDatabase.h>
#import <EXUpdates/EXUpdatesFileDownloader.h>
#import <EXUpdates/EXUpdatesModule.h>
#import <EXUpdates/EXUpdatesRemoteAppLoader.h>
#import <EXUpdates/EXUpdatesService.h>
#import <EXUpdates/EXUpdatesUpdate.h>

#if __has_include(<EXUpdates/EXUpdates-Swift.h>)
#import <EXUpdates/EXUpdates-Swift.h>
#else
#import "EXUpdates-Swift.h"
#endif

@interface EXUpdatesModule ()

@property (nonatomic, weak) id<EXUpdatesModuleInterface> updatesService;

@end

/**
 * Exported module which provides to the JS runtime information about the currently running update
 * and updates state, along with methods to check for and download new updates, reload with the
 * newest downloaded update applied, and read/clear native log entries.
 *
 * Communicates with the updates hub (EXUpdatesAppController in most apps, EXAppLoaderExpoUpdates in
 * Expo Go and legacy standalone apps) via EXUpdatesService, an internal module which is overridden
 * by EXUpdatesBinding, a scoped module, in Expo Go.
 */
@implementation EXUpdatesModule

EX_EXPORT_MODULE(ExpoUpdates);

- (void)setModuleRegistry:(EXModuleRegistry *)moduleRegistry
{
  _updatesService = [moduleRegistry getModuleImplementingProtocol:@protocol(EXUpdatesModuleInterface)];
}

- (NSDictionary *)constantsToExport
{
  NSString *releaseChannel = _updatesService.config.releaseChannel;
  NSString *channel = _updatesService.config.requestHeaders[@"expo-channel-name"] ?: @"";
  NSString *runtimeVersion = _updatesService.config.runtimeVersion ?: @"";
  NSNumber *isMissingRuntimeVersion = @(_updatesService.config.isMissingRuntimeVersion);
  
  if (!_updatesService.isStarted) {
    return @{
      @"isEnabled": @(NO),
      @"isEmbeddedLaunch": @(NO),
      @"isMissingRuntimeVersion": isMissingRuntimeVersion,
      @"releaseChannel": releaseChannel,
      @"runtimeVersion": runtimeVersion,
      @"channel": channel
    };
  }
  EXUpdatesUpdate *launchedUpdate = _updatesService.launchedUpdate;
  if (!launchedUpdate) {
    return @{
      @"isEnabled": @(NO),
      @"isEmbeddedLaunch": @(NO),
      @"isMissingRuntimeVersion": isMissingRuntimeVersion,
      @"releaseChannel": releaseChannel,
      @"runtimeVersion": runtimeVersion,
      @"channel": channel
    };
  }

  long long commitTime = [@(floor([launchedUpdate.commitTime timeIntervalSince1970] * 1000)) longLongValue];
  
  return @{
    @"isEnabled": @(YES),
    @"isEmbeddedLaunch": @(_updatesService.isEmbeddedLaunch),
    @"isUsingEmbeddedAssets": @(_updatesService.isUsingEmbeddedAssets),
    @"updateId": launchedUpdate.updateId.UUIDString ?: @"",
    @"manifest": launchedUpdate.manifest.rawManifestJSON ?: @{},
    @"localAssets": _updatesService.assetFilesMap ?: @{},
    @"isEmergencyLaunch": @(_updatesService.isEmergencyLaunch),
    @"isMissingRuntimeVersion": isMissingRuntimeVersion,
    @"releaseChannel": releaseChannel,
    @"runtimeVersion": runtimeVersion,
    @"channel": channel,
    @"commitTime": @(commitTime)
  };
}

EX_EXPORT_METHOD_AS(reload,
                    reloadAsync:(EXPromiseResolveBlock)resolve
                         reject:(EXPromiseRejectBlock)reject)
{
  if (!_updatesService.config.isEnabled) {
    reject(@"ERR_UPDATES_DISABLED", @"You cannot reload when expo-updates is not enabled.", nil);
    return;
  }
  if (!_updatesService.canRelaunch) {
    reject(@"ERR_UPDATES_DISABLED", @"The updates module controller has not been properly initialized. If you're in development mode, you cannot use this method. Otherwise, make sure you have called [[EXUpdatesAppController sharedInstance] start].", nil);
    return;
  }

  [_updatesService requestRelaunchWithCompletion:^(BOOL success) {
    if (success) {
      resolve(nil);
    } else {
      reject(@"ERR_UPDATES_RELOAD", @"Could not reload application. Ensure you have set the `bridge` property of EXUpdatesAppController.", nil);
    }
  }];
}

EX_EXPORT_METHOD_AS(checkForUpdateAsync,
                    checkForUpdateAsync:(EXPromiseResolveBlock)resolve
                                 reject:(EXPromiseRejectBlock)reject)
{
  if (!_updatesService.config.isEnabled) {
    reject(@"ERR_UPDATES_DISABLED", @"You cannot check for updates when expo-updates is not enabled.", nil);
    return;
  }
  if (!_updatesService.isStarted) {
    reject(@"ERR_UPDATES_DISABLED", @"The updates module controller has not been properly initialized. If you're in development mode, you cannot check for updates. Otherwise, make sure you have called [[EXUpdatesAppController sharedInstance] start].", nil);
    return;
  }

  __block NSDictionary *extraHeaders;
  dispatch_sync(_updatesService.database.databaseQueue, ^{
    extraHeaders = [EXUpdatesFileDownloader extraHeadersWithDatabase:self->_updatesService.database
                                                              config:self->_updatesService.config
                                                      launchedUpdate:self->_updatesService.launchedUpdate
                                                      embeddedUpdate:self->_updatesService.embeddedUpdate];
  });

  EXUpdatesFileDownloader *fileDownloader = [[EXUpdatesFileDownloader alloc] initWithUpdatesConfig:_updatesService.config];
  [fileDownloader downloadManifestFromURL:_updatesService.config.updateUrl
                             withDatabase:_updatesService.database
                             extraHeaders:extraHeaders
                             successBlock:^(EXUpdatesUpdate *update) {
    EXUpdatesUpdate *launchedUpdate = self->_updatesService.launchedUpdate;
    EXUpdatesSelectionPolicy *selectionPolicy = self->_updatesService.selectionPolicy;
    if ([selectionPolicy shouldLoadNewUpdate:update withLaunchedUpdate:launchedUpdate filters:update.manifestFilters]) {
      resolve(@{
        @"isAvailable": @(YES),
        @"manifest": update.manifest.rawManifestJSON
      });
    } else {
      resolve(@{
        @"isAvailable": @(NO)
      });
    }
  } errorBlock:^(NSError *error) {
    reject(@"ERR_UPDATES_CHECK", error.localizedDescription, error);
  }];
}

EX_EXPORT_METHOD_AS(readLogEntriesAsync,
                     readLogEntriesAsync:(NSNumber *)maxAge
                                 resolve:(EXPromiseResolveBlock)resolve
                                  reject:(EXPromiseRejectBlock)reject)
{
  EXUpdatesLogReader *reader = [EXUpdatesLogReader new];
  NSError *error = nil;
  // maxAge is in milliseconds, convert to seconds to compute NSDate
  NSTimeInterval age = [maxAge intValue] / 1000;
  NSDate *epoch = [NSDate dateWithTimeIntervalSinceNow:-age];
  NSArray<NSDictionary *> *entries = [reader getLogEntriesNewerThan:epoch error:&error];
  if (error != nil) {
    reject(@"ERR_UPDATES_READ_LOGS", [error localizedDescription], error);
  } else {
    resolve(entries);
  }
}

EX_EXPORT_METHOD_AS(clearLogEntriesAsync,
                     clearLogEntriesAsync:(EXPromiseResolveBlock)resolve
                                   reject:(EXPromiseRejectBlock)reject)
{
  EXUpdatesLogReader *reader = [EXUpdatesLogReader new];
  [reader purgeLogEntriesOlderThan:[NSDate date] completion:^(NSError *error) {
    if (error) {
      reject(@"ERR_UPDATES_READ_LOGS", [error localizedDescription], error);
    } else {
      resolve(nil);
    }
  }];
}

EX_EXPORT_METHOD_AS(fetchUpdateAsync,
                    fetchUpdateAsync:(EXPromiseResolveBlock)resolve
                              reject:(EXPromiseRejectBlock)reject)
{
  if (!_updatesService.config.isEnabled) {
    reject(@"ERR_UPDATES_DISABLED", @"You cannot fetch updates when expo-updates is not enabled.", nil);
    return;
  }
  if (!_updatesService.isStarted) {
    reject(@"ERR_UPDATES_DISABLED", @"The updates module controller has not been properly initialized. If you're in development mode, you cannot fetch updates. Otherwise, make sure you have called [[EXUpdatesAppController sharedInstance] start].", nil);
    return;
  }

  EXUpdatesRemoteAppLoader *remoteAppLoader = [[EXUpdatesRemoteAppLoader alloc] initWithConfig:_updatesService.config database:_updatesService.database directory:_updatesService.directory launchedUpdate:_updatesService.launchedUpdate completionQueue:self.methodQueue];
  [remoteAppLoader loadUpdateFromUrl:_updatesService.config.updateUrl onManifest:^BOOL(EXUpdatesUpdate * _Nonnull update) {
    return [self->_updatesService.selectionPolicy shouldLoadNewUpdate:update withLaunchedUpdate:self->_updatesService.launchedUpdate filters:update.manifestFilters];
  } asset:^(EXUpdatesAsset *asset, NSUInteger successfulAssetCount, NSUInteger failedAssetCount, NSUInteger totalAssetCount) {
    // do nothing for now
  } success:^(EXUpdatesUpdate * _Nullable update) {
    if (update) {
      [self->_updatesService resetSelectionPolicy];
      resolve(@{
        @"isNew": @(YES),
        @"manifest": update.manifest.rawManifestJSON
      });
    } else {
      resolve(@{
        @"isNew": @(NO)
      });
    }
  } error:^(NSError * _Nonnull error) {
    reject(@"ERR_UPDATES_FETCH", @"Failed to download new update", error);
  }];
}

@end

File: packages/expo-updates/ios/EXUpdates/EXUpdatesModule.swift
// Copyright 2019 650 Industries. All rights reserved.

// swiftlint:disable closure_body_length
// swiftlint:disable function_body_length

import ExpoModulesCore

/**
 * Exported module which provides to the JS runtime information about the currently running update
 * and updates state, along with methods to check for and download new updates, reload with the
 * newest downloaded update applied, and read/clear native log entries.
 *
 * Communicates with the updates hub (EXUpdatesAppController in most apps, EXAppLoaderExpoUpdates in
 * Expo Go and legacy standalone apps) via EXUpdatesService, an internal module which is overridden
 * by EXUpdatesBinding, a scoped module, in Expo Go.
 */
public class EXUpdatesModule: Module {
  private let updatesService: EXUpdatesModuleInterface?
  private let methodQueue = DispatchQueue(label: "expo.modules.EXUpdatesQueue")

  public required init(appContext: AppContext) {
    updatesService = appContext.legacyModule(implementing: EXUpdatesModuleInterface.self)
    super.init(appContext: appContext)
  }

  public func definition() -> ModuleDefinition {
    Name("EXUpdates")

    Constants {
      let releaseChannel = updatesService?.config.releaseChannel
      let channel = updatesService?.config.requestHeaders["expo-channel-name"] ?? ""
      let runtimeVersion = updatesService?.config.runtimeVersion ?? ""
      let isMissingRuntimeVersion = updatesService?.config.isMissingRuntimeVersion()

      guard let updatesService = updatesService,
        updatesService.isStarted,
        let launchedUpdate = updatesService.launchedUpdate else {
        return [
          "isEnabled": false,
          "isEmbeddedLaunch": false,
          "isMissingRuntimeVersion": isMissingRuntimeVersion,
          "releaseChannel": releaseChannel,
          "runtimeVersion": runtimeVersion,
          "channel": channel
        ]
      }

      let commitTime = UInt64(floor(launchedUpdate.commitTime.timeIntervalSince1970 * 1000))
      return [
        "isEnabled": true,
        "isEmbeddedLaunch": updatesService.isEmbeddedLaunch,
        "isUsingEmbeddedAssets": updatesService.isUsingEmbeddedAssets,
        "updateId": launchedUpdate.updateId.uuidString,
        "manifest": launchedUpdate.manifest.rawManifestJSON(),
        "localAssets": updatesService.assetFilesMap ?? [:],
        "isEmergencyLaunch": updatesService.isEmergencyLaunch,
        "isMissingRuntimeVersion": isMissingRuntimeVersion,
        "releaseChannel": releaseChannel,
        "runtimeVersion": runtimeVersion,
        "channel": channel,
        "commitTime": commitTime,
        "nativeDebug": EXUpdatesUtils.isNativeDebuggingEnabled()
      ]
    }

    AsyncFunction("reloadAsync") { (promise: Promise) in
      guard let updatesService = updatesService, updatesService.config.isEnabled else {
        throw UpdatesDisabledException()
      }
      guard updatesService.canRelaunch else {
        throw UpdatesNotInitializedException()
      }
      updatesService.requestRelaunch { success in
        if success {
          promise.resolve(nil)
        } else {
          promise.reject(UpdatesReloadException())
        }
      }
    }

    AsyncFunction("checkForUpdateAsync") { (promise: Promise) in
      guard let updatesService = updatesService, updatesService.config.isEnabled else {
        throw UpdatesDisabledException()
      }
      guard updatesService.isStarted else {
        throw UpdatesNotInitializedException()
      }

      var extraHeaders: [String: Any] = [:]
      updatesService.database.databaseQueue.sync {
        extraHeaders = EXUpdatesFileDownloader.extraHeaders(
          withDatabase: updatesService.database,
          config: updatesService.config,
          launchedUpdate: updatesService.launchedUpdate,
          embeddedUpdate: updatesService.embeddedUpdate
        )
      }

      let fileDownloader = EXUpdatesFileDownloader(config: updatesService.config)
      fileDownloader.downloadManifest(
        // swiftlint:disable:next force_unwrapping
        fromURL: updatesService.config.updateUrl!,
        withDatabase: updatesService.database,
        extraHeaders: extraHeaders
      ) { update in
        let launchedUpdate = updatesService.launchedUpdate
        let selectionPolicy = updatesService.selectionPolicy
        if selectionPolicy.shouldLoadNewUpdate(update, withLaunchedUpdate: launchedUpdate, filters: update.manifestFilters) {
          promise.resolve([
            "isAvailable": true,
            "manifest": update.manifest.rawManifestJSON()
          ])
        }
      } errorBlock: { error in
        promise.reject("ERR_UPDATES_CHECK", error.localizedDescription)
      }
    }

    AsyncFunction("readLogEntriesAsync") { (maxAge: Int) -> [[String: Any]] in
      // maxAge is in milliseconds, convert to seconds
      do {
        return try UpdatesLogReader().getLogEntries(newerThan: Date(timeIntervalSinceNow: TimeInterval(-1 * (maxAge / 1000))))
      } catch {
        throw Exception(name: "ERR_UPDATES_READ_LOGS", description: error.localizedDescription)
      }
    }

    AsyncFunction("clearLogEntriesAsync") { (promise: Promise) in
      UpdatesLogReader().purgeLogEntries(olderThan: Date()) { error in
        guard let error = error else {
          promise.resolve(nil)
          return
        }
        promise.reject("ERR_UPDATES_READ_LOGS", error.localizedDescription)
      }
    }

    AsyncFunction("fetchUpdateAsync") { (promise: Promise) in
      guard let updatesService = updatesService, updatesService.config.isEnabled else {
        throw UpdatesDisabledException()
      }
      guard updatesService.isStarted else {
        throw UpdatesNotInitializedException()
      }

      let remoteAppLoader = EXUpdatesRemoteAppLoader(
        config: updatesService.config,
        database: updatesService.database,
        directory: updatesService.directory,
        launchedUpdate: updatesService.launchedUpdate,
        completionQueue: methodQueue
      )
      remoteAppLoader.loadUpdate(
        // swiftlint:disable:next force_unwrapping
        fromURL: updatesService.config.updateUrl!
      ) { update in
        return updatesService.selectionPolicy.shouldLoadNewUpdate(
          update,
          withLaunchedUpdate: updatesService.launchedUpdate,
          filters: update.manifestFilters
        )
      } asset: { _, _, _, _ in
        // do nothing for now
      } success: { update in
        if let update = update {
          updatesService.resetSelectionPolicy()
          promise.resolve([
            "isNew": true,
            "manifest": update.manifest.rawManifestJSON()
          ])
        } else {
          promise.resolve([
            "isNew": false
          ])
        }
      } error: { error in
        promise.reject("ERR_UPDATES_FETCH", "Failed to download new update: \(error.localizedDescription)")
      }
    }
  }
}
